sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpPipelineBuilder.java

```diff
@@ -90,20 +131,411 @@ public HttpPipelineBuilder httpClient(HttpClient httpClient) {
+            afterInstrumentation)) {
+            return this;
+        }
+
+        throw LOGGER.logThrowableAsError(new IllegalStateException("Should not reach here."));
+    }
+
+    private boolean attemptToAdd(HttpPipelinePolicy policy, String policyName, boolean before,
+        LinkedList<HttpPipelinePolicy> beforePolicies, HttpPipelinePolicy keyPolicy,
+        LinkedList<HttpPipelinePolicy> afterPolicies) {
+        int position = 0;
+        for (HttpPipelinePolicy toCheck : beforePolicies) {
+            if (toCheck.getName().equalsIgnoreCase(policyName)) {
+                if (before) {
+                    beforePolicies.add(position, policy);
+                } else {
+                    if (position == beforePolicies.size() - 1) {
+                        beforePolicies.add(policy);
+                    } else {
+                        beforePolicies.add(position + 1, policy);
+                    }
+                }
+                return true;
+            }
+
+            position++;
+        }
+
+        if (keyPolicy != null && keyPolicy.getName().equalsIgnoreCase(policyName)) {
+            if (before) {
+                beforePolicies.add(policy);
+            } else {
+                afterPolicies.push(policy);
+            }
+            return true;
+        }
+
+        position = 0;
+        for (HttpPipelinePolicy toCheck : afterPolicies) {
+            if (toCheck.getName().equalsIgnoreCase(policyName)) {
+                if (before) {
+                    afterPolicies.add(position, policy);
+                } else {
+                    if (position == afterPolicies.size() - 1) {
+                        afterPolicies.add(policy);
+                    } else {
+                        afterPolicies.add(position + 1, policy);
+                    }
+                }
+                return true;
+            }
+
+            position++;
+        }
+
+        return false;
+    }
+
+    /**
+     * Replaces an existing policy with the given {@link HttpPipelinePolicy#getName()} with the provided {@code policy}.
+     * <p>
+     * If the policy with the given name is not found an {@link NoSuchElementException} will be thrown.
+     * <p>
+     * If the name of the {@code policy} is any of {@code "redirect"}, {@code "retry"}, {@code "credential"},
+     * {@code "logging"}, or {@code "telemetry"} an {@link IllegalStateException} will be thrown. These names are
+     * reserved for the set methods that configure the respective policies.
+     *
+     * @param policy The policy to set.
+     * @return The updated HttpPipelineBuilder object.
+     * @throws NullPointerException If {@code policy} is null.
+     * @throws NoSuchElementException If a policy with the given name is not found.
+     * @throws IllegalStateException If the name of the {@code policy} is reserved.
+     */
+    public HttpPipelineBuilder setPolicy(HttpPipelinePolicy policy) {
```

Comment:
If we need a given policy to already be present in the pipeline, we should rename this `replacePolicy()` instead of `setPolicy()`; "set" implies that you can add if regardless of it being present.

Reply:
Good idea, I'll rename this to `replacePolicy`.